client/r/rhawc/R/rclientClass.R

```diff
@@ -0,0 +1,158 @@
+#' r client functions
+#'
+#' pretty please work
+#'
+#' @import httr
+#' @import plyr
+#' @export
+HawcClient <- function(baseUrl){
+
+  # local environment variables
+  thisEnv <- environment()
+
+  # persistent "state"
+  root_url <- baseUrl
+  #   token <- ""
+
+  checkIsAuthenticated <- function(){
+    if((nchar(get("token", thisEnv))==0)){
+      stop('Authentication required')
+    }
+    else if (get("token", thisEnv) == 'Unable to log in with provided credentials.') {
+      stop('Unable to log in with provided credentials.')
+    }
+  }
+  get_ = function(url) {
+    response = GET(url, add_headers(Authorization=paste('Token', get("token", thisEnv),sep=' '),
+                                    'Content-Type'="application/json"))
+    if (response$status_code >= 400 & response$status_code < 500) {
+      stop("An exception occured in the HAWC client module") }
+
+    else if (response$status_code == 500) {
+      stop("An exception occured on the HAWC server") }
+
+    else {
+      response.list = content(response)
+      response.list = lapply(response.list, lapply, function(x)ifelse(is.null(x), NA, x))
+      response.df = rbind.fill(lapply(response.list, as.data.frame))
+      return(response.df) }
+  }
+  post_ = function(url, ...) {
+    response = POST(url, add_headers(Authorization=paste('Token', get("token", thisEnv),sep=' '),
+                                     'Content-Type'="application/json"), ...)
+    if (response$status_code >= 400 & response$status_code < 500) {
+      stop("An exception occured in the HAWC client module") }
+    else if (response$status_code == 500) {
+      stop("An exception occured on the HAWC server") }
+    else {
+      response.list = content(response)
+      response.list = lapply(response.list, lapply, function(x)ifelse(is.null(x), NA, x))
+      response.df = rbind.fill(lapply(response.list, as.data.frame))
+      return(response.df) }
+  }
+  # the methods which will be exposed by the client
+  me <- list(
+    authenticate = function(username, password){
+      response = POST(paste0(root_url, "/user/api/token-auth/"),
+                      body = list(username = username, password = password), encode = "json")
+      token = (content(response))[[1]]
+      assign("token", token, thisEnv)
+    },
+    doSomethingAuthenticated = function(){
+      checkIsAuthenticated()
+      return(TRUE)
+    },
+    lit_import_hero = function(assessment_id, title, description, ids) {
+      url = paste0(root_url, "/lit/api/search/")
```

Comment:
instead of paste0 everywhere, can we use glue instead? seems much more readable?

https://github.com/tidyverse/glue

```r
glue::glue("{root_url}/lit/api/search/")
```